Source/Core/DolphinQt2/FIFOWindow.cpp

```diff
@@ -0,0 +1,359 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/FIFOWindow.h"
+
+#include <QCheckBox>
+#include <QDialogButtonBox>
+#include <QFileDialog>
+#include <QGridLayout>
+#include <QGroupBox>
+#include <QLabel>
+#include <QMessageBox>
+#include <QPushButton>
+#include <QSpinBox>
+
+#include "Core/FifoPlayer/FifoDataFile.h"
+#include "Core/FifoPlayer/FifoPlaybackAnalyzer.h"
+#include "Core/FifoPlayer/FifoPlayer.h"
+#include "Core/FifoPlayer/FifoRecorder.h"
+
+#include "DolphinQt2/QtUtils/QueueOnObject.h"
+#include "DolphinQt2/Settings.h"
+
+FIFOWindow::FIFOWindow(QWidget* parent) : QDialog(parent)
+{
+  setWindowTitle(tr("FIFO Player"));
+
+  CreateWidgets();
+  ConnectWidgets();
+
+  UpdateInfo();
+
+  connect(this, &FIFOWindow::EmulationStarted, this, &FIFOWindow::OnEmulationStarted);
+  connect(this, &FIFOWindow::EmulationStopped, this, &FIFOWindow::OnEmulationStopped);
+  connect(this, &FIFOWindow::EmulationStopped, this, &FIFOWindow::UpdateInfo);
+
+  connect(&Settings::Instance(), &Settings::FIFORecordingDone, this, &FIFOWindow::OnRecordingDone);
```

Comment:
These shouldn't be signals on Settings (since they're only used internally), and shouldn't even be signals: they should be called directly, like `QueueOnObject(this, [this] { OnFIFOLoaded(); });`.

This will require a change to FIFOPlayer to take std::function callbacks, instead of just bare function pointers.

Reply:
That's preferable, yes. But I would rather not touch it.

Reply:
Why don't you want to touch it? It'd be three lines changed (swap `std::function` in for the function pointer types) that improve core and make a more understandable class here, versus this larger tangle of signals and connections.

Actually, after looking at it more, two of the callbacks could live/be connected in Settings, since they're related to the static FifoPlayer instance, but even then fixing FifoPlayer to take `std::function` would make that code simpler.